### ISSUE TITLE
acctupdatetime

### ISSUE THREAD

**Issue by iamenderst** [lirantal/daloradius#249]
Fresh Ubuntu 20.04 install 
freeradius 3.0.20+dfsg-3ubuntu0.1
daloRADIUS 1.3

Log going crazy with the following:
`
Thu Jul 28 18:42:33 2022 : ERROR: (0) sql: ERROR: rlm_sql_mysql: ERROR 1054 (Unknown column 'acctupdatetime' in 'field list'): 42S22
`
I see old issues with missing columns in  the radacct table, must still be an issue?

**Comment by lirantal** [comment]
Looks like your database doesn't have a compatible table layout. As in, the mismatch between FreeRADIUS's expected schema and what's actually in the database.

**Comment by PDS-GH** [comment]
_You're saying this is a difference in table layout with Freeradius v2 vs Freeradius v3?_

This is more than likely. There even is a db schema difference between Freeradius 3.0.17 and 3.0.20 (at least on Rocky Linux 8).

I strongly suggest analyzing the db schema provided in the freeradius package:
    /etc/raddb/mods-config/sql/main/mysql/schema.sql
An compare it with your db schema.

**Comment by MiguelVis** [comment]
Yes, unfortunately there are differences between Freeradius v2 and v3 db schemes.

I think we should try to manage this in Daloradius but it would not be a simple task.

Anyway it's a must if we want Daloradius working with FR 3.

Any help regarding this (telling failing options, etc.) will be highly appreciated.

**Comment by iamenderst** [comment]
What can I provide that would help sort this out? To me seems like I've done a basic install, no options and at the end of the install process I import our DB from an older install. Then configure a new/test nas and test. That's when I start seeing those log entries.

**Comment by PDS-GH** [comment]
As I mentioned, you have to compare your current DB schema (coming from FR2) with the target DB schema (FR3).

For example when I migrated from FR 3.0.17 to FR 3.0.20:

Current DB schema of FR 3.0.17 for table radacct:

```
MariaDB [radius]> show create table radacct;
| radacct | CREATE TABLE `radacct` (
  `radacctid` bigint(21) NOT NULL AUTO_INCREMENT,
  `acctsessionid` varchar(64) NOT NULL DEFAULT '',
  `acctuniqueid` varchar(32) NOT NULL DEFAULT '',
  `username` varchar(64) NOT NULL DEFAULT '',
  `realm` varchar(64) DEFAULT '',
  `nasipaddress` varchar(15) NOT NULL DEFAULT '',
  `nasportid` varchar(15) DEFAULT NULL,
  `nasporttype` varchar(32) DEFAULT NULL,
  `acctstarttime` datetime DEFAULT NULL,
  `acctupdatetime` datetime DEFAULT NULL,
  `acctstoptime` datetime DEFAULT NULL,
  `acctinterval` int(12) DEFAULT NULL,
  `acctsessiontime` int(12) unsigned DEFAULT NULL,
  `acctauthentic` varchar(32) DEFAULT NULL,
  `connectinfo_start` varchar(50) DEFAULT NULL,
  `connectinfo_stop` varchar(50) DEFAULT NULL,
  `acctinputoctets` bigint(20) DEFAULT NULL,
  `acctoutputoctets` bigint(20) DEFAULT NULL,
  `calledstationid` varchar(50) NOT NULL DEFAULT '',
  `callingstationid` varchar(50) NOT NULL DEFAULT '',
  `acctterminatecause` varchar(32) NOT NULL DEFAULT '',
  `servicetype` varchar(32) DEFAULT NULL,
  `framedprotocol` varchar(32) DEFAULT NULL,
  `framedipaddress` varchar(15) NOT NULL DEFAULT '',
  PRIMARY KEY (`radacctid`),
  UNIQUE KEY `acctuniqueid` (`acctuniqueid`),
  KEY `username` (`username`),
  KEY `framedipaddress` (`framedipaddress`),
  KEY `acctsessionid` (`acctsessionid`),
  KEY `acctsessiontime` (`acctsessiontime`),
  KEY `acctstarttime` (`acctstarttime`),
  KEY `acctinterval` (`acctinterval`),
  KEY `acctstoptime` (`acctstoptime`),
  KEY `nasipaddress` (`nasipaddress`)
) ENGINE=InnoDB AUTO_INCREMENT=1667026 DEFAULT CHARSET=latin1 |
```

When checking against target DB schema of FR 3.0.20 for table radacct:

```
more /etc/raddb/mods-config/sql/main/mysql/schema.sql
CREATE TABLE IF NOT EXISTS radacct (
  radacctid bigint(21) NOT NULL auto_increment,
  acctsessionid varchar(64) NOT NULL default '',
  acctuniqueid varchar(32) NOT NULL default '',
  username varchar(64) NOT NULL default '',
  realm varchar(64) default '',
  nasipaddress varchar(15) NOT NULL default '',
  nasportid varchar(15) default NULL,
  nasporttype varchar(32) default NULL,
  acctstarttime datetime NULL default NULL,
  acctupdatetime datetime NULL default NULL,
  acctstoptime datetime NULL default NULL,
  acctinterval int(12) default NULL,
  acctsessiontime int(12) unsigned default NULL,
  acctauthentic varchar(32) default NULL,
  connectinfo_start varchar(50) default NULL,
  connectinfo_stop varchar(50) default NULL,
  acctinputoctets bigint(20) default NULL,
  acctoutputoctets bigint(20) default NULL,
  calledstationid varchar(50) NOT NULL default '',
  callingstationid varchar(50) NOT NULL default '',
  acctterminatecause varchar(32) NOT NULL default '',
  servicetype varchar(32) default NULL,
  framedprotocol varchar(32) default NULL,
  framedipaddress varchar(15) NOT NULL default '',
  framedipv6address varchar(45) NOT NULL default '',
  framedipv6prefix varchar(45) NOT NULL default '',
  framedinterfaceid varchar(44) NOT NULL default '',
  delegatedipv6prefix varchar(45) NOT NULL default '',
  PRIMARY KEY (radacctid),
  UNIQUE KEY acctuniqueid (acctuniqueid),
  KEY username (username),
  KEY framedipaddress (framedipaddress),
  KEY framedipv6address (framedipv6address),
  KEY framedipv6prefix (framedipv6prefix),
  KEY framedinterfaceid (framedinterfaceid),
  KEY delegatedipv6prefix (delegatedipv6prefix),
  KEY acctsessionid (acctsessionid),
  KEY acctsessiontime (acctsessiontime),
  KEY acctstarttime (acctstarttime),
  KEY acctinterval (acctinterval),
  KEY acctstoptime (acctstoptime),
  KEY nasipaddress (nasipaddress)
) ENGINE = INNODB;
```

You can notice the differences regarding keys and missing fields, which can be solved - in my case - by doing the following ALTER TABLE:

```
ALTER TABLE radacct
    ADD (framedipv6address varchar(45) NOT NULL default '',
        framedipv6prefix varchar(45) NOT NULL default '',
        framedinterfaceid varchar(44) NOT NULL default '',
        delegatedipv6prefix varchar(45) NOT NULL default '',
        KEY framedipv6address (framedipv6address),
        KEY framedipv6prefix (framedipv6prefix),
        KEY framedinterfaceid (framedinterfaceid),
        KEY delegatedipv6prefix (delegatedipv6prefix));
```

This is just an example for radacct and for my FR version. You have to do the exercise for all you tables.

In a nutshell:
- Do `show create table <table>` for all your current radius tables.
- Compare with table structure provided by _your_ target FR version (be it 3.0.20 or another).
- Generate ALTER TABLE to catch up differences for each tables.

That should fix your table schema issues.

**Comment by iamenderst** [comment]
For me it's:
`ALTER TABLE radacct  
   ADD (acctupdatetime datetime NULL default NULL,
       acctinterval int(12) default NULL,
       framedipv6address varchar(45) NOT NULL default '',
       framedipv6prefix varchar(45) NOT NULL default '',
       framedinterfaceid varchar(44) NOT NULL default '',
       delegatedipv6prefix varchar(45) NOT NULL default '',
       KEY framedipv6address (framedipv6address),
       KEY framedipv6prefix (framedipv6prefix),
       KEY framedinterfaceid (framedinterfaceid),
       KEY delegatedipv6prefix (delegatedipv6prefix),
       KEY acctinterval (acctinterval));`

Will be able to test Monday. Thanks for the guidance.

**Comment by PDS-GH** [comment]
You should also check the other tables.

For example this is the other differences that I had between FR 3.0.17 and FR 3.0.20

```
ALTER TABLE radusergroup ADD id int(11) unsigned NOT NULL AUTO_INCREMENT PRIMARY KEY FIRST;

ALTER TABLE radpostauth ADD KEY username (username(32));

ALTER TABLE radippool MODIFY COLUMN pool_key varchar(30) NOT NULL default '';
```

**Comment by iamenderst** [comment]
For me, everything I had to touch from freeradius v2.x to freeradius v3.0.20:
```
ALTER TABLE radacct
   ADD (acctupdatetime datetime NULL default NULL,
   acctinterval int(12) default NULL,
   framedipv6address varchar(45) NOT NULL default '',
   framedipv6prefix varchar(45) NOT NULL default '',
   framedinterfaceid varchar(44) NOT NULL default '',
   delegatedipv6prefix varchar(45) NOT NULL default '',
   KEY framedipv6address (framedipv6address),
   KEY framedipv6prefix (framedipv6prefix),
   KEY framedinterfaceid (framedinterfaceid),
   KEY delegatedipv6prefix (delegatedipv6prefix),
   KEY acctinterval (acctinterval));
   
ALTER TABLE radusergroup
   add (id int(11) unsigned NOT NULL auto_increment,
   PRIMARY KEY  (id));
   
ALTER TABLE radpostauth ADD KEY username (username(32));
```
I didn't have table radippool.
For me this issue is closed but I see this has been mentioned in #253, if the devs want to close this issue, I'm ok with that.

**Comment by elonden** [comment]
> For me, everything I had to touch from freeradius v2.x to freeradius v3.0.20:
> 
> ```
> ALTER TABLE radacct
>    ADD (acctupdatetime datetime NULL default NULL,
>    acctinterval int(12) default NULL,
>    framedipv6address varchar(45) NOT NULL default '',
>    framedipv6prefix varchar(45) NOT NULL default '',
>    framedinterfaceid varchar(44) NOT NULL default '',
>    delegatedipv6prefix varchar(45) NOT NULL default '',
>    KEY framedipv6address (framedipv6address),
>    KEY framedipv6prefix (framedipv6prefix),
>    KEY framedinterfaceid (framedinterfaceid),
>    KEY delegatedipv6prefix (delegatedipv6prefix),
>    KEY acctinterval (acctinterval));
>    
> ALTER TABLE radusergroup
>    add (id int(11) unsigned NOT NULL auto_increment,
>    PRIMARY KEY  (id));
>    
> ALTER TABLE radpostauth ADD KEY username (username(32));
> ```
> 
> I didn't have table radippool. For me this issue is closed but I see this has been mentioned in #253, if the devs want to close this issue, I'm ok with that.

Just as an fyi. This also worked for me.

**Comment by DutchGalah** [comment]
Still a current issue for new installs with a master taken from Jan2023 BTW
